### ISSUE TITLE
Unable to identify distant binary on remote machine

### ISSUE THREAD

**Issue by luanzeba** [chipsenkbeil/distant.nvim#69]
It appears that distant.nvim is not able to find the distant binary on the remote machine.

I'm using distant.nvim with a [GitHub Codespace](https://github.com/features/codespaces).

## Reproduction steps

https://user-images.githubusercontent.com/20481048/174425429-a5d47b69-3131-4455-bb24-0b7033683faa.mp4

More details:
```shell
# Create the codespace using the GitHub CLI
gh cs create -r luanzeba/eventz
# SSH using port 1234
gh cs ssh --server-port 1234
# installed distant using cargo on the remote machine (codespace)
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
cargo install distant
```

In my local machine, starting neovim and running `:DistantLaunch localhost ssh.port=1234 ssh.user=codespace` leads to the following error:
```
callback error
stack traceback:
^I[C]: in ?
^I[C]: in function 'pcall'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:81: in function 'load'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:67: in function 'launch'
^I...config/nvim/plugged/distant.nvim/lua/distant/command.lua:177: in function 'launch'
^I[string ":lua"]:1: in main chunk
caused by: Spawning distant failed [???]: bash: distant: command not found
callback error
stack traceback:
^I[C]: in ?
^I[C]: in function 'pcall'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:81: in function 'load'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:67: in function 'launch'
^I...config/nvim/plugged/distant.nvim/lua/distant/command.lua:177: in function 'launch'
^I[string ":lua"]:1: in main chunk
caused by: Spawning distant failed [???]: bash: distant: command not found
```

Running distant launch in ssh mode works fine `:DistantLaunch localhost mode=ssh ssh.port=1234 ssh.user=codespace`

## Software details
* remote machine OS: Ubuntu 20.04.4 LTS (GNU/Linux 5.4.0-1074-azure x86_64)
* cargo on remote machine: `cargo 1.61.0 (a028ae4 2022-04-29)`
* distant on remote machine: `distant 0.16.4`
* `which distant` on remote: `/home/codespace/.cargo/bin/distant`
* distant on local MacOS: `distant 0.16.4`
* neovim version: ` NVIM v0.7.0`

**Comment by chipsenkbeil** [comment]
We try to detect the path to distant, but it doesn't work that well right now. Launching distant isn't able to take advantage of your path, so it would need to be under the default path like `/usr/bin` or maybe `/usr/local/bin`. You can specify the path for the distant binary on the remote machine as an option via `:DistantLaunch distant.bin=/home/codespace/.cargo/bin/distant`.

**Comment by luanzeba** [comment]
Thanks for the quick response @chipsenkbeil!
Unfortunately, both alternatives have led to a different error `Connection refused (os error 61)`:
1. Specifying the path
`:DistantLaunch localhost mode=ssh ssh.port=1234 ssh.user=codespace distant.bin=/home/codespace/.cargo/bin/distant `
2. Symlinking the binary
```
sudo ln -sfv /home/codespace/.cargo/bin/distant /usr/bin

:DistantLaunch localhost mode=ssh ssh.port=1234 ssh.user=codespace
```


Both lead to this error
```
callback error
stack traceback:
^I[C]: in ?
^I[C]: in function 'pcall'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:81: in function 'load'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:67: in function 'launch'
^I...config/nvim/plugged/distant.nvim/lua/distant/command.lua:177: in function 'launch'
^I[string ":lua"]:1: in main chunk
caused by: Connection refused (os error 61)
callback error
stack traceback:
^I[C]: in ?
^I[C]: in function 'pcall'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:81: in function 'load'
^I.../nvim/plugged/distant.nvim/lua/distant/editor/launch.lua:67: in function 'launch'
^I...config/nvim/plugged/distant.nvim/lua/distant/command.lua:177: in function 'launch'
^I[string ":lua"]:1: in main chunk
caused by: Connection refused (os error 61)
```

Some basic googling says that my network connection may not be configured properly or there is a firewall blocking the connection.. I don't _believe_ I have any firewalls on the remote machine. These are the services running that I could find
```
service --status-all
 [ ? ]  binfmt-support
 [ - ]  dbus
 [ ? ]  hwclock.sh
 [ - ]  procps
 [ - ]  rsync
 [ + ]  ssh
 [ - ]  x11-common
```
Let me know if you have any ideas of steps I can take to further troubleshoot it

**Comment by chipsenkbeil** [comment]
Check and see if the distant server is running on your remote machine after you get a connection refused. It may be that the server died and so nothing is listening on the port you described. I would also check what port the distant server bound to on your remote machine.

If you set the log file for the server, you should see what it bound to. Also, given that this is a codespace, they may have restricted the ports that are accessible from the outside. SSH's 22 is a staple, so if codespace supports ssh then it makes sense for that port to be open.

If you know that a certain port is open for TCP, you can specify a port for distant to listen on (versus it trying for an available one) using `distant.args = "--port 1234"`. By default, `distant` listens on the next available port in the range 8080 to 8099.

**Comment by luanzeba** [comment]
Thanks for all the guidance, that did it!
> If you know that a certain port is open for TCP, you can specify a port for distant to listen on (versus it trying for an available one) using `distant.args="--port 1234"`

Alternatively, I was also able to open port 8080 in the codespace and just use that since it's the distant default as you pointed out.

I think we should be able to close out this issue. That being said, I ran into a separate problem when using the distant binary instead of ssh. I'm not able to open files using `:DistantOpen`. It seems to just hang there. I was tailing the logs but nothing came up.
I recorded the behavior below and showed that it still works in ssh mode (sorry about the small font). By the way, after I recorded it, I also tried with `--log-level=debug` and `--log-level=trace` and nothing came up either.

https://user-images.githubusercontent.com/20481048/174680948-c986bfe7-64f9-49ca-81a5-bf444a5edf92.mp4

Let me know if I should open a separate issue with these details
